keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,752 @@
+
+Deployment
+- The current behavior remains unchanged when the DeploymentPodReplacementPolicy feature gate is disabled or PodReplacementPolicy is nil.
+  The `.status.terminatingReplicas` field should be 0 in that case.
+- Add a test wrapper for any relevant tests, to ensure that they are run with all possible PodReplacementPolicy values correctly.
+  The relevant tests are those that expect some behavior on Pod deletion, and are affected by this change.
+- New unit tests should be added for any new helper functions.
+- Test that the status is computed correctly.
+- Test feature gate enablement and disablement.
+
+
+The core packages (with their unit test coverage) which are going to be modified during the implementation:
+- `k8s.io/kubernetes/pkg/apis/apps/v1`: `9 December 2023` - `71.4%` <!--(extension of Deployment and ReplicaSet API)-->
+- `k8s.io/kubernetes/pkg/apis/apps/validation`: `9 December 2023` - `92.3%` <!--(validation of the PodReplacementPolicy)-->
+- `k8s.io/kubernetes/pkg/controller/deployment`: `9 December 2023` - `61.7%` <!--(implementation of the PodReplacementPolicy)-->
+- `k8s.io/kubernetes/pkg/controller/deployment/util`: `9 December 2023` - `50.1%` <!--(implementation of the PodReplacementPolicy)-->
+- `k8s.io/kubernetes/pkg/controller/replicaset`: `9 December 2023` - `78.9%` <!--(implementation of the counting of the TerminatingReplicas)-->
+- `k8s.io/kubernetes/pkg/controller`: `9 December 2023` - `71.2%` <!--(creating or updating the utility functions in controller_utils.go)-->
+
+##### Integration tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+Integration tests covering:
+
+<!-- test/integration/replicaset/replicaset_test.go -->
+ReplicaSet
+- The current behavior remains unchanged when the DeploymentPodReplacementPolicy feature gate is disabled.
+- Add a new test that correctly counts `.status.terminatingReplicas` when the DeploymentPodReplacementPolicy feature gate is enabled.
+
+- <test>: <link to test coverage>
+
+<!-- test/integration/deployment/deployment_test.go -->
+Deployment
+- The current behavior remains unchanged when the DeploymentPodReplacementPolicy feature gate is disabled or PodReplacementPolicy is nil.
+- Add a test wrapper for any relevant tests, to ensure that they are run with all possible PodReplacementPolicy values correctly.
+  The relevant tests are those that expect some behavior on Pod deletion, and are affected by this change.
+- Add new tests that observe rollout and scaling transitions for all possible PodReplacementPolicy values and
+  ensure that `.status.terminatingReplicas` is correctly counted when the DeploymentPodReplacementPolicy feature gate is enabled.
+
+- <test>: <link to test coverage>
+
+##### e2e tests
```

Comment:
As I worked on the promotion to Beta for this feature in Job controller, there were a couple of caveats we encountered along the way when wanting to force the pods into terminating state for some amount of time in order to assert status, so I'll post the PR here https://github.com/kubernetes/kubernetes/pull/121491, as it might save somebody's time

Reply:
I expect the deployment tests to be a bit different, but it might come handy. Thanks!